lib/pbench/server/auth/auth.py

```diff
@@ -101,3 +107,61 @@ def verify_auth(auth_token):
                 e,
             )
         return None
+
+    @staticmethod
+    def verify_third_party_token(
+        auth_token: str, oidc_client: OpenIDClient, tokeninfo_endpoint: str = None
+    ) -> bool:
+        """
+        Verify a token provided to the Pbench server which was obtained from a
+        third party identity provider.
+        Args:
+            auth_token: Token to authenticate
+            oidc_client: OIDC client to call to authenticate the token
+            tokeninfo_endpoint: Optional tokeninfo_endpoint to validate
+                                tokens online in case offline verification
+                                results in some exception.
+        Returns:
+            True if the verification succeed else False
+        """
+        # Verify auth token validity
+        identity_provider_pubkey = oidc_client.get_oidc_public_key(auth_token)
+        try:
+            oidc_client.token_introspect_offline(
+                token=auth_token,
+                key=identity_provider_pubkey,
+                audience=oidc_client.client_id,
+                options={
+                    "verify_signature": True,
+                    "verify_aud": True,
+                    "verify_exp": True,
+                },
+            )
+            return True
+        except (
+            jwt.ExpiredSignatureError,
+            jwt.InvalidTokenError,
+            jwt.InvalidAudienceError,
+        ):
+            Auth.logger.error("oidc token verification failed")
+        except Exception as e:
+            Auth.logger.exception(
+                "Unexpected exception occurred while verifying the auth token {}: {}",
+                auth_token,
+                e,
+            )
+            if not tokeninfo_endpoint:
+                Auth.logger.warning("Can not perform oidc online token verification")
+            else:
+                try:
+                    token_payload = oidc_client.token_introspect_online(
+                        token=auth_token, token_info_uri=tokeninfo_endpoint
+                    )
+                    if "aud" not in token_payload:
+                        return False
+                    elif oidc_client.client_id in token_payload["aud"]:
+                        # If our client is an intended audience then only we accept the token
+                        return True
+                except OpenIDClientError:
+                    return False
+        return False
```

Comment:
Consider recasting this code like this:
```
        except (
            jwt.ExpiredSignatureError,
            jwt.InvalidTokenError,
            jwt.InvalidAudienceError,
        ):
            Auth.logger.error("oidc token verification failed")
            return False
        except Exception as e:
            Auth.logger.exception(
                "Unexpected exception occurred while verifying the auth token {}: {}",
                auth_token,
                e,
            )

        if not tokeninfo_endpoint:
            Auth.logger.warning("Can not perform oidc online token verification")
            return False

        try:
            token_payload = oidc_client.token_introspect_online(
                token=auth_token, token_info_uri=tokeninfo_endpoint
            )
        except OpenIDClientError:
            return False

        return "aud" in token_payload and oidc_client.client_id in token_payload["aud"]
```
That is, if control escapes the `try` block (and doesn't return from the function), then we will have logged the unexpected exception and we proceed (and everything is de-indented by one level).  If the necessary endpoint is unavailable, then we return immediately, otherwise we proceed (and everything is de-indented by another level!)  Then I've moved the unrelated stuff out of the final `try` block (so it's de-indented yet another level!!).  And, finally, we return The Answer.